File: app/src/main/java/com/addedfooddelivery_user/chat/adpter/ChatAdapter.java
package com.addedfooddelivery_user.chat.adpter;

import android.content.ContentValues;
import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.recyclerview.widget.RecyclerView;
import com.addedfooddelivery_user.R;

import java.util.ArrayList;
import java.util.List;



public class ChatAdapter extends RecyclerView.Adapter {

    private int VIEW_SEND = 0;
    private int VIEW_RECIVED = 1;

    private Context context;
    private List<String> chatObjects;

    public ChatAdapter(Context context, ArrayList<String> chatObjects) {
        this.context = context;
        this.chatObjects = chatObjects;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == VIEW_SEND) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_msg_outgoing, parent, false);
            return new SendViewHolder(v);
        } else if (viewType == VIEW_RECIVED) {
            View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_msg_incoming, parent, false);
            return new RecievedViewHolder(v);
        }
        return null;
    }


    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {
        if (holder instanceof SendViewHolder) {
            final SendViewHolder sendViewHolder = (SendViewHolder) holder;

           /* sendViewHolder.messageChatTv.setText(chatObjects.get(position).getAsString("message"));
            sendViewHolder.timeChatTv.setText(SmartUtils.getTime(chatObjects.get(position).getAsString("message_time")));*/

        } else if (holder instanceof RecievedViewHolder) {
            final RecievedViewHolder recievedViewHolder = (RecievedViewHolder) holder;
           /* recievedViewHolder.timeChatTv.setText(SmartUtils.getTime(chatObjects.get(position).getAsString("message_time")));*/
        }
    }

    @Override
    public int getItemCount() {
        if (chatObjects != null) {
            return chatObjects.size();
        } else {
            return 0;
        }
    }

    @Override
    public int getItemViewType(int position) {
        int result = 0;

        if (chatObjects.get(position).equalsIgnoreCase("1")) {
            result = VIEW_SEND;
        } else {
            result = VIEW_RECIVED;
        }
        return result;
    }


    private class SendViewHolder extends RecyclerView.ViewHolder {

        private SendViewHolder(View itemView) {
            super(itemView);


        }
    }

    private class RecievedViewHolder extends RecyclerView.ViewHolder {

        private RecievedViewHolder(View itemView) {
            super(itemView);


        }
    }
}

File: PubNubChatDemo/app/src/main/java/com/example/pubnubchatdemo/chatAdapter.kt
package com.example.pubnubchatdemo

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.pubnubchatdemo.dataclass.ChatMessages

class chatAdapter(var array: ArrayList<ChatMessages>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    //    var MSG = 0
//    var IMG = 1
    var SENDER = 0
    var RECIEVER = 1
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): RecyclerView.ViewHolder {
        return when (viewType) {
            SENDER -> {
                return MessageViewHolder1(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.chat_row_layout_sender_msg, parent, false)
                )
            }
            RECIEVER -> {
                MessageViewHolder2(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.chat_row_layout_reciever_msg, parent, false)
                )
            }
            else -> {
                return MessageViewHolder1(
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.chat_row_layout_sender_msg, parent, false)
                )
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is MessageViewHolder1 -> holder.bind(array[position], position)
            is MessageViewHolder2 -> holder.bind(array[position], position)
        }

    }

    override fun getItemCount(): Int {
        return array.size
    }

    inner class MessageViewHolder1(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var txtSendMsg: TextView = itemView.findViewById(R.id.txtSendMsg)
        var imgRight: ImageView = itemView.findViewById(R.id.imgRight)


        fun bind(data: ChatMessages, position: Int) {
            txtSendMsg.text = data.msg
            imgRight.LoadImg(data.url.toString())
            if (data.is_image == true) {
                imgRight.visibility = View.VISIBLE
                txtSendMsg.visibility = View.GONE
            } else {
                imgRight.visibility = View.GONE
                txtSendMsg.visibility = View.VISIBLE
            }

        }

    }

    inner class MessageViewHolder2(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var txtRecieveMsg: TextView = itemView.findViewById(R.id.txtRecieveMsg)
        var imgLeft: ImageView = itemView.findViewById(R.id.imgLeft)

        fun bind(data: ChatMessages, position: Int) {
            imgLeft.LoadImg(data.url.toString())
            txtRecieveMsg.text = data.msg

            if (data.is_image == true) {

                imgLeft.visibility = View.VISIBLE
                txtRecieveMsg.visibility = View.GONE
            } else {
                imgLeft.visibility = View.GONE
                txtRecieveMsg.visibility = View.VISIBLE
            }

        }

    }

    override fun getItemViewType(position: Int): Int {
        val type: Int = if (array[position].sender == Appconstants.UUID) {
            SENDER
        } else {
            RECIEVER
        }
        return type
    }
}

fun ImageView.LoadImg(url: String) {
    Glide.with(this.context)
        .load(url)
        .placeholder(R.mipmap.ic_launcher)
        .into(this)

}
